server/src/internalClusterTest/java/org/elasticsearch/snapshots/SnapshotsServiceIT.java

```diff
@@ -121,4 +126,42 @@ public void testSnapshotDeletionFailureShouldBeLogged() throws Exception {
             deleteRepository("test-repo");
         }
     }
+
+    public void testRerouteWhenShardSnapshotsCompleted() throws Exception {
+        final var repoName = randomIdentifier();
+        createRepository(repoName, "mock");
+        internalCluster().ensureAtLeastNumDataNodes(1);
+        final var originalNode = internalCluster().startDataOnlyNode();
+
+        final var indexName = randomIdentifier();
+        createIndexWithContent(
+            indexName,
+            indexSettings(1, 0).put(IndexMetadata.INDEX_ROUTING_REQUIRE_GROUP_PREFIX + "._name", originalNode).build()
+        );
+
+        final var snapshotFuture = startFullSnapshotBlockedOnDataNode(randomIdentifier(), repoName, originalNode);
+
+        // Use allocation filtering to push the shard to a new node, but it will not do so yet because of the ongoing snapshot.
+        updateIndexSettings(
+            Settings.builder()
+                .putNull(IndexMetadata.INDEX_ROUTING_REQUIRE_GROUP_PREFIX + "._name")
+                .put(IndexMetadata.INDEX_ROUTING_EXCLUDE_GROUP_PREFIX + "._name", originalNode)
+        );
+
+        final var shardMovedListener = ClusterServiceUtils.addTemporaryStateListener(
+            internalCluster().getCurrentMasterNodeInstance(ClusterService.class),
+            state -> {
+                final var primaryShard = state.routingTable().index(indexName).shard(0).primaryShard();
+                return primaryShard.started() && originalNode.equals(state.nodes().get(primaryShard.currentNodeId()).getName()) == false;
+            }
+        );
+        assertFalse(shardMovedListener.isDone());
+
+        unblockAllDataNodes(repoName);
+        assertEquals(SnapshotState.SUCCESS, snapshotFuture.get(10, TimeUnit.SECONDS).getSnapshotInfo().state());
+
+        // Now that the snapshot completed the shard should move to its new location.
+        safeAwait(shardMovedListener);
+        ensureGreen(indexName);
```

Comment:
Above ensureGreen checks for no relocating and no initializing shards.
But does it guarantee relocation even started due to the exclude rule? 
I wonder if this test could silently pass if index is green because it never moved

Reply:
No, because `shardMovedListener` completed successfully.